feat: serialize synchronize and simulateTx calls by the pxe via SerialQueue

just-mitch opened the pull request:
Serialize calls by the PXE to:
- synchronize with the aztec node
- simulate transactions
by use of a SerialQueue

Synchronization is greedy, meaning it will sync as much as possible while its job is processing.

just-mitch commented on yarn-project/foundation/src/fifo/memory_fifo.ts:
```diff
@@ -91,9 +91,9 @@ export class MemoryFifo<T> {
   /**
    * Process items from the queue using a provided handler function.
-   * The function iterates over items in the queue, invoking the handler for each item until the queue is empty or flushing.
+   * The function iterates over items in the queue, invoking the handler for each item until the queue is empty and flushing.
```
See the `get` function above. the `process` loop here only breaks when we get `null`, which only happens `if (this.items.length === 0 && this.flushing)`

just-mitch commented on yarn-project/pxe/src/synchronizer/synchronizer.test.ts:
```diff
@@ -31,7 +33,8 @@ describe('Synchronizer', () => {
     database = new KVPxeDatabase(await AztecLmdbStore.create(EthAddress.random()));
-    synchronizer = new TestSynchronizer(aztecNode, database);
+    jobQueue = new SerialQueue();
+    synchronizer = new TestSynchronizer(aztecNode, database, jobQueue);
```
the job queue is only used when `synchronizer.start` is used. This doesn't happen in this test, nor does it happen in the pxe_service tests.

PhilWindle commented on yarn-project/pxe/src/synchronizer/synchronizer.ts:
```diff
@@ -42,17 +43,19 @@ export class Synchronizer {
-          // No note processor needs to catch up. We continue with the normal flow.
-          await this.work(limit, retryInterval);
-        }
+        await this.jobQueue.put(async () => {
```
We will need to change this sightly. Currently, the `work` functions sleep if there is nothing more to do.

So we want the `work` functions to return indicating if there is more work to do. Then inside the job we loop until there is no more work to be done. Finally, at the outer layer (outside of the job queue) we sleep for the `retryInterval`. 

Does that sound like it works?

PhilWindle commented on yarn-project/pxe/src/pxe_service/pxe_service.ts:
```diff
@@ -336,18 +345,22 @@ export class PXEService implements PXE {
-      await this.#simulatePublicCalls(tx);
-    }
-    this.log.info(`Executed local simulation for ${await tx.getTxHash()}`);
+    // we need to ensure that we are not overlapping with the synchronizer
```
We will also need to enqueue the code within `viewTx` below.

Commit:
Serialize calls by the PXE to:
- synchronize with the aztec node
- simulate transactions
by use of a SerialQueue

Synchronization is greedy, meaning it will sync as much as possible while its job is processing.
